libnavigation-base/src/main/java/com/mapbox/navigation/base/route/Router.kt

```diff
@@ -3,18 +3,44 @@ package com.mapbox.navigation.base.route
 import com.mapbox.api.directions.v5.models.DirectionsRoute
 import com.mapbox.api.directions.v5.models.RouteOptions
 
+/**
+ * Router provides API to fetch route and cancel route-fetching request.
+ */
 interface Router {
 
+    /**
+     * Fetch route based on [RouteOptions]
+     *
+     * @param routeOptions RouteOptions
+     * @param callback Callback that should be handled
+     */
     fun getRoute(
         routeOptions: RouteOptions,
         callback: Callback
     )
 
+    /**
+     * Interrupt route-fetching request
+     */
     fun cancel()
 
+    /**
+     * Callback for Router fetching
+     */
     interface Callback {
+
+        /**
+         * List of [DirectionsRoute]
+         *
+         * @param routes List<DirectionsRoute> the most relevant has index 0, second is 1 and etc.
+         * Has at least one Route
+         */
         fun onResponse(routes: List<DirectionsRoute>)
 
+        /**
+         * @param throwable Throwable safety error.
+         * Called on I/O either when network error occurred or no one Route has been found.
```

Comment:
I think error callback is called on the main thread as well, isn't it?

Reply:
with courutines yes, but I think we shouldn't guarantee it

Reply:
I think we should. All our callbacks, if not required by the internal design, should return on the main thread.

Reply:
Also now _it is_ returned on the main thread, so this doc would be kinda confusing :)

Reply:
could you clarify why we should guarantee that callback will be invoked in Main thread?

Reply:
That's how ours and most of the other Android libraries operate. We should always return on the main thread unless there's a strong and documented performance reason not to do so. It makes consuming the library that much easier for everyone.

Reply:
yes, some libraries work as you mentions, but not most. 
For sample OkHttp library doesn't return request-result to main thread(if you don't set it explicitly). Rx doens't as well, you should handle it on your side. Vision(Mapbox) doesn't return result to Main thread(or do it not always) as well because as a developer of SDK you don't know what user will do with callback result(not always result parsed in Main Thread).  Main thing that I want to say we shouldn't be owners of customer intensions, we don't know how user would use result from callback. 

IMO The most democracy way is provide to user way to set Handler(or in our case Courutione Dispatcher) where he want to receive result

Reply:
Users can always post to the worker thread themselves if that's what's desired. Making input and output methods main thread only (excluding strongly incentivized use-cases) would most of all keep our SDK consistent and we wouldn't have to worry about making every single class thread-safe when all it does is lightweight tasks. If we start treating callbacks differently, one returning on the main thread, one on the worker, one on the OkHttp thread it would just be consuming and unmaintainable.

We can always consider exposing alternative methods that accept handlers for return types, but as I mentioned above, users can already do that themselves.

Reply:
User also can reschedule callback to main thread, this works on both ways. 
I think this issue is system design one. I didn't note that we worry where and how user calls methods of SDK(we don't make them synchronized or/and don't reschedule calls to main thread).

Reply:
@LukasPaczos ptal. I removed note regarding thread. Comment now about error type